zfit/core/basepdf.py

```diff
@@ -0,0 +1,1097 @@
+        if self.n_dims and self._norm_range is not None:
+            if not self.n_dims == self._norm_range.n_dims:
+                raise ValueError("norm_range n_dims {} does not match dist.n_dims {}"
+                                 "".format(self._norm_range.n_dims, self.n_dims))
+        else:
+            self.n_dims = self.n_dims_from_limits(norm_range)
+        try:
+            yield self.norm_range  # or None, not needed
+        finally:
+            self.set_norm_range(old_norm_range)
+
+    @property
+    def norm_range(self) -> typing.Union[Range, None]:
+        """Return the current normalization range
+
+        Returns:
+            Range or None: The current normalization range
+
+        """
+        return self._norm_range
+
+    def set_norm_range(self, norm_range: typing.Union[Range, None]):
+        """Fix the normalization range to a certain value. Use with caution!
+
+        It is, in general, better to use either the explicit `norm_range` argument when calling
+        a function or the `temp_norm_range` context manager to set a normalization range for a
+        limited amount of code.
+
+        Args:
+            norm_range ():
+
+        """
+        self._norm_range = convert_to_range(norm_range, dims=Range.FULL)
+        return self
+
+    def _check_input_norm_range(self, norm_range, dims, caller_name="",
+                                none_is_error=False) -> typing.Union[Range, bool]:
+        """If `norm_range` is None, take `self.norm_range`. Convert to :py:class:`Range`
+
+        Args:
+            norm_range (None or Range compatible):
+            dims (tuple(int)):
+            caller_name (str): name of the calling function. Used for exception message.
+            none_is_error (bool): if both `norm_range` and `self.norm_range` are None, the default
+                value is `False` (meaning: no range specified-> no normalization to be done). If
+                this is set to true, two `None` will raise a Value error.
+
+        Returns:
+            Union[Range, False]:
+
+        """
+        if norm_range is None:
+            if self.norm_range is None:
+                if none_is_error:
+                    raise ValueError("Normalization range `norm_range` has to be specified, either by"
+                                     "\na) explicitly passing it to the function {name}"
+                                     "\nb) using the temp_norm_range context manager to temporary set "
+                                     "a default normalization range. Currently, both are None/False."
+                                     "".format(name=caller_name))
+                else:
+                    norm_range = False
+            else:
+                norm_range = self.norm_range
+        return convert_to_range(limits=norm_range, dims=dims)
+
+    @property
+    def dims(self) -> typing.Tuple[int]:
+        """Return the dimensions (a tuple of integers) of the function.
```

Comment:
This one is a bit unclear... What is it its aim?

Reply:
Basically just to return the axis. So given some n-dim limits, return (0, 1, 2,...) or whatever they have been specified as. Would you prefer 'axis'?

Reply:
So, this just returns the tuple of axes? Then let's rename this to `axis`, although if we can remove it let's remove it.

Reply:
Yeah true, in the current status it is not needed. The idea was that you have a pdf with the axis (0, 3), one with (1, 2), but it's really not needed. Removed it.